### ISSUE TITLE
Filtering/querying (deep) relationships - a proposal for a syntax

### ISSUE THREAD

**Issue by cmeeren** [json-api/json-api#1445]
(Previously posted at [discuss.jsonapi.com](https://discuss.jsonapi.org/t/filtering-querying-deep-relationships-a-proposal-for-a-syntax/1746).)

JSON:API currently lacks a standard for filtering/querying data included through relationships, as opposed to GraphQL which has this firmly specified. (Related issue: #1272)

Consider the following API:
* `GET /authors` to get all resources of type `author`
* Resource `author` has relationship `articles` which is to-many `article`
* Resource `article` has attribute `isDraft: boolean`

Now consider the following request:
* "Give me all authors who have draft articles, and include their draft articles".

There are two filters here:
* The filter on the author list (filtering on a related resource's attribute)
* The filter on the included articles

Here is a syntax that might be able to confer the necessary distinction:

```
GET /authors
  ?filter[articles.isDraft]=true
  &filter[articles][isDraft]=true
  &include=articles
```

The meaning of the two filters would be:
* `filter[articles.isDraft]=true` means "only return authors who have articles with `isDraft=true`"
* `filter[articles][isDraft]=true` means "for any included articles from the `articles` relationship, only return those with `isDraft=true`"

In other words:
* The first (of two) pair of brackets indicate **what** to filter (if not present, filter top-level data)
* The second (or only) pair of brackets indicate **how** to filter, using dot notation for filtering on related resources

As far as I can see, this syntax is general enough to be nested artibrarily on any level. For example, `filter[articles.comments][author.name]=John&include=articles.comments` means "for each article, include only those comments whose author has `name=John`"

---

**Update:** It occurred to me that this proposal can also work with pagination of relationships:

```
GET /authors
  ?include=articles
  &page[articles][offset]=10
  &page[articles][limit]=5
```

The above would return all authors and include a subset of each of their articles according to the pagination parameters.

Furthermore, it can work for sparse fieldsets:

```
GET /authors
  ?include=articles.comments.author
  &fields[articles.comments.author][author]=name
  &fields[articles.comments.author][anonymousUser]=firstName
```

In the above request, the top-level (i.e. primary data) authors would have all fields, while the included authors for each comment (at the end of the include hierarchy), would either have the field `name` (if type is `author`) or `firstName` (if type is `anonymousUser`).

---

Does this make sense? Could this syntax be a clear, general, and consistent way to support filters on nested relationships? Could this be part of the JSON-API spec, or a recommendation?

**Comment by dgeb** [comment]
Opionated usage of reserved, but unspecified, query params such as `filter` will be best handled by extensions, which are coming to v1.1. See #1435 for details as well as the WIP https://github.com/json-api/json-api/tree/extension-negotiation branch.

**Comment by gabesullice** [comment]
> JSON:API currently lacks a standard for filtering/querying data included through relationships, as opposed to GraphQL which has this firmly specified.

Really? TIL! Do you happen to have a link to the documentation to the filtering standard (or equivalent) for GraphQL?

**Comment by cmeeren** [comment]
> Do you happen to have a link to the documentation to the filtering standard (or equivalent) for GraphQL?

Any basic GraphQL tutorial should suffice. :) I have only briefly experimented with GraphQL, but the whole querying syntax is built around providing this kind of capability.

**Comment by cmeeren** [comment]
It just occurred to me that this proposal can also work with pagination:

```
GET /authors
  ?include=articles
  &page[articles][offset]=10
  &page[articles][limit]=5
```

The above would return all authors and include a subset of each of their articles according to the pagination parameters.

**Comment by cmeeren** [comment]
Furthermore, it can work for sparse fieldsets:

```
GET /authors
  ?include=articles.comments.author
  &fields[articles.comments.author][author]=name
  &fields[articles.comments.author][anonymousUser]=name
```

In the above request, the top-level (i.e. primary data) authors would have all fields, while the included authors for each comment (at the end of the include hierarchy), would either have the field `name` (if type is `author`) or `firstName` (if type is `anonymousUser`).

**Comment by cmeeren** [comment]
I see that this issue addresses the "deep querying" part of #1272.

**Comment by javiertury** [comment]
It would be great if filtering deep relations was also compatible with operators. For reference here are some (non-nested) operators proposals:
https://discuss.jsonapi.org/t/share-propose-a-filtering-strategy/257/4
https://discuss.jsonapi.org/t/share-propose-a-filtering-strategy/257/23

Edit:

In #1470 I propose some alternatives to make both features compatible.

**Comment by UkonnRa** [comment]
Deep relationships' pagination is an important missing part for JSONAPI, because sometime the relationship is too large to fetch in one operation. Any update?

**Comment by UkonnRa** [comment]
In most cases the solution is simple, but there is an corner case when querying items in an array, which is hard (confusing) even using Mongo.

**Question**

Say the author/articles model above, I use Mongo as an example:

```js
db.authors.insertMany([
    {name: "author-1", articles: [{name: "article-1-1", isDrafted: true}, {name: "article-1-2", isDrafted: true}]},
    {name: "author-2", articles: [{name: "article-2-1", isDrafted: false}, {name: "article-2-2", isDrafted: true}]},
    {name: "author-3", articles: [{name: "article-3-1", isDrafted: false}, {name: "article-3-2", isDrafted: false}]},
]);
```

Here I have two situations:

1. Find all authors who has (**at least one**) drafted article
2. Find all authors whose articles are **all drafted**

**Current Solution**

All of the discussion above are about the first situation, which matches Mongo's assumption: [if any of the element matches
the query, then it passes](https://docs.mongodb.com/manual/tutorial/query-arrays/).

**Thoughts**

In Mongo, the two situations can be written as:

```js
// Authors who **has any of** drafted articles
db.authors.find({ "articles.isDrafted": { $eq: true } });

// Authors whose articles are **all** drafted (Authors who **does not has any of** un-drafted articles)
db.authors.find({ "articles.isDrafted": { $not: { $eq: false } } });
```

You can see, there is a nested query, which cannot present using URL clearly, some of the solutions like:

* `filter[articles.isDrafted][not:eq]=false` --> `not:` prefix makes the predication opposite
* `filter[articles.isDrafted][not]={ $eq: false }` --> the value of `not` operator should be another filter, which may reference new syntax - the JSON based filter
* `filter[articles.isDrafted][not]=filter[articles.isDrafted][eq]=false` --> if you think the above is not _URL-pure_

But finally, I still think that `filter` part is very complex, see the Mongo query and you can see how complex querying a nested object with different situations, so I think just use the Mongo syntax like: `filter[articles.isDrafted]={ $not: { $eq: false } }` is URL based syntax is not suitable for your query. **No other query language is as muture as Mongo query**

**-- Updated --**

What's more, the URL query cannot handle some complex query, like:

```js
db.authors.find({ $or: [ { name: "author-1" }, { "articles.name": "article-3-1" } ] });
```

In this situation, I prefer just using `filter[]=<raw-query>` like `filter[]={ $or: [ { name: "author-1" }, { "articles.name": "article-3-1" } ] }` to handle this situation

**Comment by cmeeren** [comment]
What you touch on here, is what a filter means for collections. That is a good point. In short, the question is: How many elements should match the filter? Any? All? None?

In most of my own implementations, I have so far always used an implicit "any" (or "contains") interpretation because it fits my use-cases.  For example, `filter[articles.isDrafted]=true` would mean "give me all authors who contain **at least one article** with `isDrafted=true`".

I have thought about, though not yet needed, an "exists" or "not exists" kind of functionality. I'm not sure it has to be more complicated than than e.g. `filter[articles[all].isDrafted]=true` or `filter[articles[none].isDrafted]=true`. The default "any"/"contains" interpretation could be written explicitly as `filter[articles[any].isDrafted]=true` or `filter[articles[contains].isDrafted]=true`. This syntax would work with arbitrarily deep relationships, since you can add these operators/modifiers to any part of the relationship hierarchy.

In other words, the `any`/`all`/`none` modifier can just be an operator on the relationship name.

This is just one suggestion of a syntax. I have not thought hard about whether it conflicts with anything else proposed here, but I don't think it does.

Note: In this issue, I have tried to propose a solution for filtering relationships. I have not tried to offer a full-fledged "dynamic-database-query-over-HTTP" syntax; this is just about inventing a clever and understandable filter parameter *naming* scheme and leaving the *values* simple. A full-fledged query-over-HTTP would need complex values, as you demonstrate. It also risks leaking implementation details (e.g. SQL-like syntax, Mongo-like syntax, etc.) and would be very hard to fully standardize.

**Comment by cmeeren** [comment]
It also strikes me that this can work with sorting:

```
GET /authors
  ?include=articles
  &sort=firstName,lastName
  &sort[articles]=-publishedAt
```

The above would sort the authors (the primary data) by the first and last names, and the members in the `article` relationship for each author would be sorted descending by the article's `publishedAt` attribute.

**Comment by jelhan** [comment]
> It also strikes me that this can work with sorting:
> 
> ```
> GET /authors
>   ?include=articles
>   &sort=firstName,lastName
>   &sort[articles]=-publishedAt
> ```
> 
> The above would sort the authors (the primary data) by the first and last names, and the members in the `article` relationship for each author would be sorted descending by the article's `publishedAt` attribute.

Major difference between filtering and sorting is that for filtering it could be achieved using a profile while sorting would require either change to base spec or an extension. Base spec defines processing rules for `sort` but not for `filter`. I would recommend to discuss both separately therefore.

**Comment by cmeeren** [comment]
> Base spec defines processing rules for `sort` but not for `filter`. I would recommend to discuss both separately therefore.

Could you elaborate what in the base spec is in conflict with the example I gave? Why can't that be a profile?

Perhaps this could be amended before v1.1 is released in its final version?

**Comment by jelhan** [comment]
> > Base spec defines processing rules for `sort` but not for `filter`. I would recommend to discuss both separately therefore.
> 
> Could you elaborate what in the base spec is in conflict with the example I gave? Why can't that be a profile?

JSON:API specification defines processing rules for the `sort` query parameter:

> An endpoint MAY support requests to sort the primary data with a `sort` query parameter. The value for sort MUST represent sort fields.
>
> https://jsonapi.org/format/1.1/#fetching-sorting

That conflicts with the proposal because `sort` is defined as a query parameter, not a [query parameter family](https://jsonapi.org/format/1.1/#query-parameters-families).

`sort[RELATIONSHIP]` would define a new query parameter. [Profiles](https://jsonapi.org/format/1.1/#profiles) are not allowed to do so. They are only allowed to define implementation semantics. [Implementation semantics](https://jsonapi.org/format/1.1/#semantics) can only define those "members, query parameters, and processing rules, [which] are reserved for implementors" explicitly by the base spec.

An [extension](https://jsonapi.org/format/1.1/#extensions) is allowed to define additional query parameters. These are called [extension-specific query parameters](https://jsonapi.org/format/1.1/#extension-query-parameters). They must be prefixed by the extension's namespace to avoid collisions with query parameters defined by a later iteration of the base spec. This would prevent `sort[RELATIONSHIP]` but would allow something like `relationship:sort[RELATIONSHIP]`. In this example `relationship` is the namespace declared by the extension.

Please note that it is a different story for `filter` because that one is defined by the base specification as a query parameter family, which should be used by implementations for filtering. The processing rules _how_ it is used for filtering are not defined by the base specification but defined as implementation specific:

> The filter [query parameter family](https://jsonapi.org/format/1.1/#query-parameters-families) is reserved for filtering data. Servers and clients SHOULD use these parameters for filtering operations.
>
> https://jsonapi.org/format/1.1/#fetching-filtering

> Perhaps this could be amended before v1.1 is released in its final version?

I'm not sure if it is possible to do so without a breaking change at all. The limitations of profiles and extensions have been carefully considered. The use case could be supported using an extension instead of a profile. I don't see a strong need why it needs to be possible using a profile.

**Comment by cmeeren** [comment]
Thanks for the clarification.

> The use case could be supported using an extension instead of a profile. I don't see a strong need why it needs to be possible using a profile.

I haven't actually read the 1.1 spec yet, so I wasn't aware these were two different concepts. (After reading the definitions of these terms, I'm still a bit confused.) In any case, if it's possible, I'm good.

**Comment by cmeeren** [comment]
@jelhan So if I focus on the `filter` part of this proposal, and perhaps `page`, but not `sort`, do you agree that this could be a valuable profile?

**Comment by jelhan** [comment]
> @jelhan So if I focus on the `filter` part of this proposal, and perhaps `page`, but not `sort`, do you agree that this could be a valuable profile?

I think both filtering and pagination profiles would be very valuable for the community. I'm also working on a draft for a filtering profile. For pagination, I'm only aware of the [Cursor Pagination profile](https://jsonapi.org/profiles/ethanresnick/cursor-pagination/).

I would recommend separating the filtering and pagination profiles. This makes it easier for others to adopt one or the other.

**Comment by cmeeren** [comment]
What kind of filtering profile are you working on? Should we coordinate our efforts? To avoid [this](https://xkcd.com/927/)...

**Comment by jelhan** [comment]
> What kind of filtering profile are you working on? Should we coordinate our efforts?

Let's do that!

How can I contact you? Tried LinkedIn but you may missed my message. Cannot send you private message on Twitter. Maybe you can ping me on one of the channels?

**Comment by cmeeren** [comment]
Please contact me at cmeeren@protonmail.com and we can take it from there. 👍
